File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<int>st;
        for(char ch : s){
            if(ch=='{' ||ch=='[' || ch=='('){
                st.push(ch);
            }else{
                if(st.empty()){
                    return false;
                }
                char c=st.top();
                
                if(c=='{'){
                    if(ch!='}'){
                        return false;
                    }else{
                        st.pop();
                    }
                }
                else if(c=='('){
                    if(ch!=')'){
                        return false;
                    }else{
                        st.pop();
                    }

                }else{
                    if(ch!=']'){
                        return false;
                    }else{
                        st.pop();
                    }
                }
            }
        }
        return st.empty();
    }
};

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character>st=new Stack<>();
        
        // (  { [  oeping ho raha hain to push otherwies pop  karo aur dekho ke ele hain ke hain 
       
       for(int i = 0 ;i < s.length() ;i++){
            char ch  = s.charAt(i);
            if(ch=='('  || ch=='{' || ch=='['){
                st.push(ch);
            }else{
                 if (st.empty()) {
                    return false;
                }
                 char veg=st.peek();
                 st.pop();
                 if(ch==')' && veg!='('){
                    return false;
                 }
                 if(ch=='}'   && veg!='{'){
                    return false;
                 }
                 if(ch==']'   && veg!='['){
                    return false;
                 }
            }
       }
       
       return st.empty();
    }
}